### PR TITLE
Fix worker policy race condition

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Fix a race condition between refreshing the map and retrieving worker info from the map.

### Why are the changes needed?

This cause a NPE where the requested entry can be `null`.

### Does this PR introduce any user facing changes?
No.

### CODE REVIEW

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 1]
```diff
@@ -67,32 +70,62 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      // When the active nodes map does not exist, the hash provider is not initialized yet.
+      // let one caller initialize the map while blocking all others.
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:89]: this begs the question: what if multiple callers want to initialize with different set of workers?
I think either we make `ConsistentHashProvider` non static and one instance per `FileSystemContext`, or the responsibility of initializing and refreshing it should go to the membership service, which monitors and maintains a consistent view of available workers.

**Comment by maobaolong** [comment]
@dbw9580 https://github.com/Alluxio/alluxio/pull/17718 Thanking about that there are some relation between these two PR, would you like to take my PR into this PR or take a look whether we can merge one of these two?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 2]
```diff
@@ -67,32 +71,64 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+          mInitLock.release();
+        }
       }
-      // update worker info if needed
-      if (mNeedUpdate.compareAndSet(true, false)) {
-        if (isWorkerInfoUpdated(workerInfos, mLastWorkerInfos)) {
-          build(workerInfos, numVirtualNodes);
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:109]: https://stackify.com/heres-how-to-calculate-elapsed-time-in-java/ 
Actually I think `Instant` may be better? You may be getting the nano time from different threads. I don't think that gives you a strictly monotonous behavior?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 3]
```diff
@@ -57,7 +61,7 @@ public WorkerLocationPolicy(int numVirtualNodes) {
   public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWorkerInfos,
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:61]: [May not need to fix in this PR]
I find the semantic very confusing when you can return an empty list here. Does it mean "there are no workers" or "there are workers but I simply don't prefer any ones"?

If there's only one method here, I prefer it throws an error when there are no workers. There's `ExceptionMessage.NO_WORKER_AVAILABLE` for that use.

If someone (like the job scheduler) really wants to tolerate that situation, I guess we just create two separate methods:
```
public List<> mustFindPreferredWorkers() { // throw when no workers available }

public List<> maybeFindPreferredWorkers() { // do not throw when no workers available }
```

Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:61]: @jja725 I see your CopyJob/MoveJob are using this method. Do you think the change in semantics can work?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 4]
```diff
@@ -67,32 +71,64 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
+          "cannot refresh hash provider with empty worker list");
+      // When the active nodes map does not exist, the hash provider is not initialized yet.
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:91]: what about just `synchronized`? That's mutex and simple.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 5]
```diff
@@ -67,32 +71,64 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:89]: i feel instead of throwing `IllegalArgumentEx` here, a checked exception is more explicit for callers (easier to catch and handle).

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 1 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:92]: I think a singleton per `FileSystemContext` sounds good

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 6]
```diff
@@ -115,16 +151,25 @@ public List<BlockWorkerInfo> getMultiple(String key, int count) {
 
     public BlockWorkerInfo get(String key, int index) {
       int hashKey = HASH_FUNCTION.hashString(format("%s%d", key, index), UTF_8).asInt();
-      Map.Entry<Integer, BlockWorkerInfo> entry =
-          mActiveNodesByConsistentHashing.ceilingEntry(hashKey);
+      NavigableMap<Integer, BlockWorkerInfo> map = mActiveNodesByConsistentHashing;
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:154]: this is the reader right? Is this locked?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 3 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:61]: I think we should separate the logic of refreshing the worker list and that of getting a preferred worker. The responsibility of refreshing the worker list should go to the membership service.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 7]
```diff
@@ -57,7 +61,7 @@ public WorkerLocationPolicy(int numVirtualNodes) {
   public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWorkerInfos,
                                                    String fileId,
                                                    int count) {
-    if (blockWorkerInfos.size() == 0) {
+    if (blockWorkerInfos.isEmpty()) {
       return ImmutableList.of();
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:65]: throw UnavailableException

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 2 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:91]: The javadoc of this `System.nanoTime()` is somewhat vague:
```
This method provides namesecond precision, but not necessarily nanosecond resolution (that is, how frequently the value changes).
...
The value returend by this method becomes meaningful only when the difference between two such values, obtained within the same instance of a Java virtual machine, is computed.
```

My interpretation is: from the same JVM, if multiple threads get `System.nanoTime()`, the result will be monotonous but possibly stale (seeing twice for the same timestamp). The monotocity is broken ONLY when the values are from different JVMs.

That said, I guess our use here is okay.

`Instant` uses `System.clockTime()` so can be impacted by changing system time. Also the extra thread safety may not be necessary (and is less performant).

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 4 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java]: per our discussin, we are going to use a lock for WW contentions. For RW contentions, mutual exclusion is done by the CAS on timestamp.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 6 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:150]: see https://github.com/Alluxio/alluxio/pull/17753#discussion_r1259239012

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 8]
```diff
@@ -57,42 +62,76 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
+      if (currentTs - lastUpdateTs > mWorkerInfoUpdateIntervalNs) {
+        // use CAS to only allow one thread to actually update the timestamp
+        boolean casUpdated = mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs);
+        // thread safety is valid provided that build() takes less than
+        // WORKER_INFO_UPDATE_INTERVAL_NS, so that before next update the current update has been
+        // finished
+        if (casUpdated) {
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:105]: Extract a method?
```
if (shouldRebuildHashRingExclusively())  { // extract a method here
    mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
    mLastWorkerInfos = workerInfos;
}

// Return true means should update and mutual exclusiveness is guaranteed
private boolean shouldRebuildHashRingExclusively() {
      long lastUpdateTs = mLastUpdatedTimestamp.get();
      long currentTs = System.nanoTime();
      if (currentTs - lastUpdateTs <= mWorkerInfoUpdateIntervalNs) {
        return false;
      }
      return mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs));
}
```
Of course your comments on thread safety mechanisms are still required.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 9]
```diff
@@ -57,42 +62,76 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
+    private final int mMaxAttempts;
+    private final long mWorkerInfoUpdateIntervalNs;
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    // Lazily initialized, can only be null before the first call to refresh
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
+    // Must use System.nanoTime to ensure monotonic increment
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:81]: more on how this semantic guarantees that monotonic increment

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 10]
```diff
@@ -57,42 +62,76 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+      maybeInitialize(workerInfos, numVirtualNodes);
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
+      if (currentTs - lastUpdateTs > mWorkerInfoUpdateIntervalNs) {
+        // use CAS to only allow one thread to actually update the timestamp
+        boolean casUpdated = mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs);
+        // thread safety is valid provided that build() takes less than
+        // WORKER_INFO_UPDATE_INTERVAL_NS, so that before next update the current update has been
+        // finished
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:104]: Add a word that even if two build() are executed, last write wins. And mention when two `build()` calls might happen.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 11]
```diff
@@ -57,42 +62,76 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+    // let one caller initialize the map while blocking all others.
+    private void maybeInitialize(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
+      if (mActiveNodesByConsistentHashing == null) {
+        synchronized (ConsistentHashProvider.class) {
+          // only one thread should reach here
+          // test again to skip re-initialization
+          if (mActiveNodesByConsistentHashing == null) {
+            mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
+            mLastWorkerInfos = workerInfos;
+            mLastUpdatedTimestamp.set(System.nanoTime());
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:127]: yea let's keep using nanoTime, don't mix millis with nanoTime()

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+public class WorkerLocationPolicyTest {
+  private static final long WORKER_LIST_TTL_MS = 100;
+  private static final String OBJECT_KEY = "/path/to/object";
+  private static final int NUM_VIRTUAL_NODES = 100;
+
+  @Test
+  public void uninitializedThrowsException() {
+    ConsistentHashProvider provider = new ConsistentHashProvider(1, WORKER_LIST_TTL_MS);
+    Assert.assertThrows(IllegalStateException.class, () -> provider.get(OBJECT_KEY, 0));
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:52]: nit: import static

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  @Test
+  public void concurrentInitialization() {
+    ConsistentHashProvider provider = new ConsistentHashProvider(1, WORKER_LIST_TTL_MS);
+    final int numThreads = Runtime.getRuntime().availableProcessors();
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:58]: Lol. I guess you don't need to depend on numCPUs here. Tests on github are using probably 1 or 2 CPUs? And that can give you unexpected results against running on you Mac. I guess it's fine to keep using maybe 4 or 8. The latch will ensure those threads are concurrent (maybe not allocated CPU but they are indeed running concurrently).

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the work and the extra test! I will take a closer look to the tests first thing in the morning.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 8 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:105]: I feel that introducing another method will only add noise to the logic.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 14]
```diff
@@ -106,25 +145,49 @@ private boolean isWorkerInfoUpdated(List<BlockWorkerInfo> workerInfoList,
...
+      }
+      Map.Entry<Integer, BlockWorkerInfo> entry = map.ceilingEntry(hashKey);
       if (entry != null) {
-        return mActiveNodesByConsistentHashing.ceilingEntry(hashKey).getValue();
+        return entry.getValue();
       } else {
-        return mActiveNodesByConsistentHashing.firstEntry().getValue();
+        Map.Entry<Integer, BlockWorkerInfo> firstEntry = map.firstEntry();
+        if (firstEntry == null) {
+          throw new IllegalStateException("Hash provider is empty");
```
[the remark this answers is not in the document]
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:171]: I don't think it's worth it to put a message to another class just to reuse some strings.

The problem with the `ExceptionMessages` approach is that the exception messages are highly context-sensitive. A message can be shared by multiple contexts today, and tomorrow you find that one of the contexts needs a message that a bit different (to add more details, etc.), so you are forced to use a non-shared message. This creates a lot of noise in commit history.

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java (hunk 12 above)
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:52]: done

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java (hunk 13 above)
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:58]: changed to a fixed value of `16`.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 10 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:104]: added in the method'd javadoc/

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 15]
```diff
@@ -106,25 +145,49 @@ private boolean isWorkerInfoUpdated(List<BlockWorkerInfo> workerInfoList,
...
+      return get(mActiveNodesByConsistentHashing, key, index);
+    }
+
+    @VisibleForTesting
+    static BlockWorkerInfo get(NavigableMap<Integer, BlockWorkerInfo> map, String key, int index) {
       int hashKey = HASH_FUNCTION.hashString(format("%s%d", key, index), UTF_8).asInt();
-      Map.Entry<Integer, BlockWorkerInfo> entry =
-          mActiveNodesByConsistentHashing.ceilingEntry(hashKey);
+      if (map == null) {
+        throw new IllegalStateException("Hash provider is not properly initialized");
```
[the remark this answers is not in the document]
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java]: moved the check to the instance method `get`, where the map is an instance field, and `IllegalStateException` is more semantically accurate.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 16]
```diff
@@ -57,42 +62,106 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+     * Used to compare with incoming worker list to skip the heavy build process if the worker
+     * list has not changed.
+     */
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    /**
+     * A map of virtual node indices to the actual workers.
+     * This is lazily initialized, can only be null before the first call to refresh.
+     */
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
```
Comment (apc999) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:99]: annotation?
```suggestion
    @GuardedBy("mActiveNodesByConsistentHashing")
    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
```

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+    // check if the worker list is one of the lists provided by the threads
+    List<BlockWorkerInfo> workerInfoListUsedByPolicy = provider.getLastWorkerInfos();
+    assertTrue(lists.contains(workerInfoListUsedByPolicy));
+    assertEquals(
+        ConsistentHashProvider.build(workerInfoListUsedByPolicy, NUM_VIRTUAL_NODES),
+        provider.getActiveNodesMap());
+  }
+
+  @Test
+  // todo(bowen): this test can be flaky if the test subject is not thread safe
```
Comment (apc999) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:97]: if this is sure to be flaky, we shall not include any known flaky test

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 18]
```diff
@@ -57,42 +62,106 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
-    if (blockWorkerInfos.size() == 0) {
+    if (blockWorkerInfos.isEmpty()) {
       return ImmutableList.of();
     }
     HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
     return HASH_PROVIDER.getMultiple(fileId, count);
   }
 
-  private static class ConsistentHashProvider {
+  static class ConsistentHashProvider {
```
Comment (apc999) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:72]: it looks like we remove `private` for testing?
if so, annotate `@VisibleForTesting`

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 19]
```diff
@@ -57,42 +62,106 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+     * Counter for how many times the map has been updated.
+     */
+    private final AtomicLong mUpdateCount = new AtomicLong(0);
+    /**
+     * The worker list which the {@link #mActiveNodesByConsistentHashing} was built from.
+     * Must kept in sync with {@link #mActiveNodesByConsistentHashing}.
+     * Used to compare with incoming worker list to skip the heavy build process if the worker
+     * list has not changed.
+     */
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
```
Comment (apc999) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:93]: typically it is tricky to have a volatile List. Instead, can we have a `private final List`, but more common to copy items into this list (or a collection in general),  at least keep this reference unchanged?

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java (hunk 17 above)
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:97]: the tricky thing about this test is that it's trying to ensure the test subject is thread safe. When it is, the test is not flaky. When it is not, then it can be flaky. I can't think of a way to reliably detect if the test is thread safe or not.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 16 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:99]: It's not clear which lock object to put in the `GuardedBy` annotation. The list is protected by a synchronized lock during initialization. After initialization, it's a volatile variable, and access is synchronized by CAS on an atomic timestamp.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 19 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java]: changed to an array list and clear/copy items when the list is updated.

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+    provider.refresh(newList, NUM_VIRTUAL_NODES);
+    assertNotEquals(newList, workerList);
+    assertEquals(workerList, provider.getLastWorkerInfos());
+    assertEquals(
+        ConsistentHashProvider.build(workerList, NUM_VIRTUAL_NODES),
+        provider.getActiveNodesMap());
+
+    // after TTL expires, refresh should change the worker list and the active nodes map
+    Thread.sleep(WORKER_LIST_TTL_MS);
+    provider.refresh(newList, NUM_VIRTUAL_NODES);
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:167]: assert the refresh count has incremented

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+      try {
+        return future.get();
+      } catch (InterruptedException interruptedException) {
+        throw new AssertionError("interrupted", interruptedException);
+      } catch (ExecutionException e) {
+        throw new AssertionError("failed to run thread", e);
+      }
+    }).collect(Collectors.toSet());
+
+    assertEquals(1, mapSet.size());
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:87]: also assert refresh counter has incremented

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 22]
```diff
@@ -57,42 +63,109 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+     * Counter for how many times the map has been updated.
+     */
+    private final AtomicLong mUpdateCount = new AtomicLong(0);
+    /**
+     * The worker list which the {@link #mActiveNodesByConsistentHashing} was built from.
+     * Must be kept in sync with {@link #mActiveNodesByConsistentHashing}.
+     * Used to compare with incoming worker list to skip the heavy build process if the worker
+     * list has not changed.
+     */
+    private final List<BlockWorkerInfo> mLastWorkerInfos = new ArrayList<>();
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:95]: i think you can keep a Set here instead of generating one on each comparison

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 23]
```diff
@@ -57,42 +63,109 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+     */
+    private void maybeInitialize(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
+      if (mActiveNodesByConsistentHashing == null) {
+        synchronized (mInitLock) {
+          // only one thread should reach here
+          // test again to skip re-initialization
+          if (mActiveNodesByConsistentHashing == null) {
+            mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
+            mLastWorkerInfos.clear();
+            mLastWorkerInfos.addAll(workerInfos);
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:160]: actually instead of clearing/adding, what about an AtomicRef to the new Set/List?

Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:160]: I just feel like `final` with a mutable data structure is a bit inconsistent. And under Dora you might have 1000+ worker nodes, which makes a list/set generation non-trivial.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 24]
```diff
@@ -57,42 +63,109 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+    /**
+     * Timestamp of the last update to {@link #mActiveNodesByConsistentHashing}.
+     * Must use System.nanoTime to ensure monotonic increment. Otherwise, updates to the worker
+     * list may be missed as the expiry based on TTL cannot be reliably determined.
+     */
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+    /**
+     * Counter for how many times the map has been updated.
+     */
+    private final AtomicLong mUpdateCount = new AtomicLong(0);
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:88]: Use LongAdder for better performance

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java [hunk 25]
```diff
@@ -57,42 +63,109 @@ public WorkerLocationPolicy(int numVirtualNodes) {
...
+     * The worker list which the {@link #mActiveNodesByConsistentHashing} was built from.
+     * Must be kept in sync with {@link #mActiveNodesByConsistentHashing}.
+     * Used to compare with incoming worker list to skip the heavy build process if the worker
+     * list has not changed.
+     */
+    private final List<BlockWorkerInfo> mLastWorkerInfos = new ArrayList<>();
+    /**
+     * A map of virtual node indices to the actual workers.
+     * This is lazily initialized, can only be null before the first call to refresh.
+     */
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:99]:
```
/**
  * Requirements:
  * 1. This hash ring is lazy initialized (cannot init in the constructor). Multiple threads may try to enter the init section and it should be only initialized once.
  * 2. This hash ring is timestamped. After the TTL expires, we need to compare the worker list with the current available workers and possibly rebuild the hash ring.
  * 3. While the hash ring is being updated, readers should see a stale hash ring without blocking.
  * 
  * Thread safety guarantees:
  * 1. At lazy-init time, mutual exclusion is provided by `synchronized(mInitLock)` and double-checking. At this stage it is guarded by `mInitLock`.
  * 2. After init, updating the hash ring is guarded by an optimistic lock using CAS(timestamp). There will be no blocking but a read may see a stale ring. At this stage it is guarded by `mLastUpdatedTimestamp`.
  */
```

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 16 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java:101]: My best try would be a comment like above. Maybe we don't need `@GuardedBy` as long as the javadoc here is clear enough.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
I think this is my last batch of comments. Most LGTM and thanks a lot for the work! It's really hard to get these code right.

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java (hunk 21 above)
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:87]: initialization does not increment update count

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 8 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java]: extracted into a method

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java (hunk 17 above)
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java]: added comments on the test flakiness

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java (hunk 22 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java]: I'll leave this to another PR.

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java [hunk 26]
```diff
@@ -94,52 +94,62 @@ public void concurrentInitialization() {
...
 
   @Test
-  // todo(bowen): this test can be flaky if the test subject is not thread safe
+  // Notes on thread safety:
+  // This test tries to ensure the test subject live up to its thread safety guarantees.
+  // 1. When the test subject is actually correctly implemented, this test will *always* pass.
+  // 2. When it is not, this test cannot reliably detect that, since it's a matter of
+  //    chance that race conditions manifest themselves. This test may appear to be flaky,
+  //    but chances are that there are thread safety issues with the test subject.
+  // To decrease the chance of false negatives, you can run this test manually multiple times
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java:103]: nit: Add a word that "In this test we run concurrent operations for 100 times to reduce the chance of a false negative."

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks for the high quality work!

**Comment by dbw9580** [comment]
alluxio-bot, merge this please.
